README.rst
Requests: HTTP for Humans
=========================

.. image:: https://img.shields.io/pypi/v/requests.svg
    :target: https://pypi.python.org/pypi/requests

.. image:: https://img.shields.io/pypi/l/requests.svg
    :target: https://pypi.python.org/pypi/requests

.. image:: https://img.shields.io/pypi/pyversions/requests.svg
    :target: https://pypi.python.org/pypi/requests

.. image:: https://codecov.io/github/requests/requests/coverage.svg?branch=master
    :target: https://codecov.io/github/requests/requests
    :alt: codecov.io

.. image:: https://img.shields.io/github/contributors/requests/requests.svg
    :target: https://github.com/requests/requests/graphs/contributors

.. image:: https://img.shields.io/badge/Say%20Thanks-!-1EAEDB.svg
    :target: https://saythanks.io/to/kennethreitz


元文書: [requests/requests](https://github.com/requests/requests), [#a3d7cf3f27e74c28ef30f01e9f2e483570ab042e](https://github.com/requests/requests/tree/a3d7cf3f27e74c28ef30f01e9f2e483570ab042e)

このドキュメントは [Requests](http://docs.python-requests.org/) のドキュメントを翻訳したものです。
翻訳の間違いなどの指摘は、[Issues](https://github.com/requests/requests-docs-jp/issues)に登録して頂けるとうれしいです。


Requests is the only *Non-GMO* HTTP library for Python, safe for human
consumption.

**Warning:** Recreational use of the Python standard library for HTTP may result in dangerous side-effects,
including: security vulnerabilities, verbose code, reinventing the wheel,
constantly reading documentation, depression, headaches, or even death.

Behold, the power of Requests:

.. code-block:: python

    >>> r = requests.get('https://api.github.com/user', auth=('user', 'pass'))
    >>> r.status_code
    200
    >>> r.headers['content-type']
    'application/json; charset=utf8'
    >>> r.encoding
    'utf-8'
    >>> r.text
    u'{"type":"User"...'
    >>> r.json()
    {u'disk_usage': 368627, u'private_gists': 484, ...}

See `the similar code, sans Requests <https://gist.github.com/973705>`_.

.. image:: https://raw.githubusercontent.com/requests/requests/master/docs/_static/requests-logo-small.png
    :target: http://docs.python-requests.org/


Requests allows you to send *organic, grass-fed* HTTP/1.1 requests, without the
need for manual labor. There's no need to manually add query strings to your
URLs, or to form-encode your POST data. Keep-alive and HTTP connection pooling
are 100% automatic, thanks to `urllib3 <https://github.com/shazow/urllib3>`_.

Besides, all the cool kids are doing it. Requests is one of the most
downloaded Python packages of all time, pulling in over 11,000,000 downloads
every month. You don't want to be left out!

Feature Support
---------------

Requests is ready for today's web.

- International Domains and URLs
- Keep-Alive & Connection Pooling
- Sessions with Cookie Persistence
- Browser-style SSL Verification
- Basic/Digest Authentication
- Elegant Key/Value Cookies
- Automatic Decompression
- Automatic Content Decoding
- Unicode Response Bodies
- Multipart File Uploads
- HTTP(S) Proxy Support
- Connection Timeouts
- Streaming Downloads
- ``.netrc`` Support
- Chunked Requests

Requests officially supports Python 2.6–2.7 & 3.3–3.7, and runs great on PyPy.

Installation
------------

To install Requests, simply:

.. code-block:: bash

    $ pip install requests
    ✨🍰✨

Satisfaction guaranteed.

Documentation
-------------

Fantastic documentation is available at http://docs.python-requests.org/, for a limited time only.


How to Contribute
-----------------

#. Check for open issues or open a fresh issue to start a discussion around a feature idea or a bug. There is a `Contributor Friendly`_ tag for issues that should be ideal for people who are not very familiar with the codebase yet.
#. Fork `the repository`_ on GitHub to start making your changes to the **master** branch (or branch off of it).
#. Write a test which shows that the bug was fixed or that the feature works as expected.
#. Send a pull request and bug the maintainer until it gets merged and published. :) Make sure to add yourself to AUTHORS_.

.. _`the repository`: http://github.com/requests/requests
.. _AUTHORS: https://github.com/requests/requests/blob/master/AUTHORS.rst
.. _Contributor Friendly: https://github.com/requests/requests/issues?direction=desc&labels=Contributor+Friendly&page=1&sort=updated&state=open


HISTORY.rst
.. :changelog:

.. Release History
   ---------------

リリース履歴
---------------

2.18.4 (2017-08-15)
+++++++++++++++++++

**改善**

.. Error messages for invalid headers now include the header name for easier debugging

- 無効なヘッダーのエラーメッセージにデバッグを容易にするヘッダー名が追加されました。

**依存**

.. We now support idna v2.6.

- idna のバージョン 2.6 をサポートしました。

2.18.3 (2017-08-02)
+++++++++++++++++++

**改善**

.. Running ``$ python -m requests.help`` now includes the installed version of idna.

- ``$ python -m requests.help`` を実行すると、インストールされているバージョンの idna が含まれるようになりました。

**バグフィックス**

- Fixed issue where Requests would raise ``ConnectionError`` instead of
  ``SSLError`` when encountering SSL problems when using urllib3 v1.22.

2.18.2 (2017-07-25)
+++++++++++++++++++

**Bugfixes**

- ``requests.help`` no longer fails on Python 2.6 due to the absence of
  ``ssl.OPENSSL_VERSION_NUMBER``.

**Dependencies**

- We now support urllib3 v1.22.

2.18.1 (2017-06-14)
+++++++++++++++++++

**Bugfixes**

- Fix an error in the packaging whereby the *.whl contained incorrect data that
  regressed the fix in v2.17.3.

2.18.0 (2017-06-14)
+++++++++++++++++++

**Improvements**

- ``Response`` is now a context manager, so can be used directly in a ``with`` statement
  without first having to be wrapped by ``contextlib.closing()``.

**Bugfixes**

- Resolve installation failure if multiprocessing is not available
- Resolve tests crash if multiprocessing is not able to determine the number of CPU cores
- Resolve error swallowing in utils set_environ generator


2.17.3 (2017-05-29)
+++++++++++++++++++

**Improvements**

- Improved ``packages`` namespace identity support, for monkeypatching libraries.


2.17.2 (2017-05-29)
+++++++++++++++++++

**Improvements**

- Improved ``packages`` namespace identity support, for monkeypatching libraries.


2.17.1 (2017-05-29)
+++++++++++++++++++

**Improvements**

- Improved ``packages`` namespace identity support, for monkeypatching libraries.


2.17.0 (2017-05-29)
+++++++++++++++++++

**Improvements**

- Removal of the 301 redirect cache. This improves thread-safety.


2.16.5 (2017-05-28)
+++++++++++++++++++

- Improvements to ``$ python -m requests.help``.

2.16.4 (2017-05-27)
+++++++++++++++++++

- Introduction of the ``$ python -m requests.help`` command, for debugging with maintainers!

2.16.3 (2017-05-27)
+++++++++++++++++++

- Further restored the ``requests.packages`` namespace for compatibility reasons.

2.16.2 (2017-05-27)
+++++++++++++++++++

- Further restored the ``requests.packages`` namespace for compatibility reasons.

No code modification (noted below) should be neccessary any longer.

2.16.1 (2017-05-27)
+++++++++++++++++++

- Restored the ``requests.packages`` namespace for compatibility reasons.
- Bugfix for ``urllib3`` version parsing.

**Note**: code that was written to import against the ``requests.packages``
namespace previously will have to import code that rests at this module-level
now.

For example::

    from requests.packages.urllib3.poolmanager import PoolManager

Will need to be re-written to be::

    from requests.packages import urllib3
    urllib3.poolmanager.PoolManager

Or, even better::

    from urllib3.poolmanager import PoolManager

2.16.0 (2017-05-26)
+++++++++++++++++++

- Unvendor ALL the things!

2.15.1 (2017-05-26)
+++++++++++++++++++

- Everyone makes mistakes.

2.15.0 (2017-05-26)
+++++++++++++++++++

**Improvements**

- Introduction of the ``Response.next`` property, for getting the next
  ``PreparedResponse`` from a redirect chain (when ``allow_redirects=False``).
- Internal refactoring of ``__version__`` module.

**Bugfixes**

- Restored once-optional parameter for ``requests.utils.get_environ_proxies()``.

2.14.2 (2017-05-10)
+++++++++++++++++++

**Bugfixes**

- Changed a less-than to an equal-to and an or in the dependency markers to
  widen compatibility with older setuptools releases.

2.14.1 (2017-05-09)
+++++++++++++++++++

**Bugfixes**

- Changed the dependency markers to widen compatibility with older pip
  releases.

2.14.0 (2017-05-09)
+++++++++++++++++++

**Improvements**


requests/exceptions.py
# -*- coding: utf-8 -*-

"""
requests.exceptions
~~~~~~~~~~~~~~~~~~~

This module contains the set of Requests' exceptions.
"""
from urllib3.exceptions import HTTPError as BaseHTTPError


class RequestException(IOError):
    """
    .. There was an ambiguous exception that occurred while handling your
       request.

    リクエストの処理中に原因不明な例外が起きた時に送出されます。
    """

    def __init__(self, *args, **kwargs):
        """Initialize RequestException with `request` and `response` objects."""
        response = kwargs.pop('response', None)
        self.response = response
        self.request = kwargs.pop('request', None)
        if (response is not None and not self.request and
                hasattr(response, 'request')):
            self.request = self.response.request
        super(RequestException, self).__init__(*args, **kwargs)


class HTTPError(RequestException):
    """
    .. An HTTP error occurred.

    HTTP エラーが起きた時に送出されます。
    """


class ConnectionError(RequestException):
    """
    .. A Connection error occurred.

    コネクションエラーが起きた時に送出されます。
    """


class ProxyError(ConnectionError):
    """A proxy error occurred."""


class SSLError(ConnectionError):
    """An SSL error occurred."""


class Timeout(RequestException):
    """
    .. The request timed out.

    リクエストがタイムアウトした時に送出されます。

    .. Catching this error will catch both
       :exc:`~requests.exceptions.ConnectTimeout` and
       :exc:`~requests.exceptions.ReadTimeout` errors.

    このエラーを受け取ると、:exc:`~requests.exceptions.ConnectTimeout` と
    :exc:`~requests.exceptions.ReadTimeout` の2つのエラーを受け取ることになります。
    """


class ConnectTimeout(ConnectionError, Timeout):
    """
    .. The request timed out while trying to connect to the remote server.

    リモートサーバーに接続しようとして、リクエストがタイムアウトした時に送出されます。

    .. Requests that produced this error are safe to retry.

    このエラーが発生したリクエストは、再度実行しても問題ありません。
    """


class ReadTimeout(Timeout):
    """
    .. The server did not send any data in the allotted amount of time.

    サーバーが設定された時間内にデータを返却しなかった場合に送出されます。
    """


class URLRequired(RequestException):
    """
    .. A valid URL is required to make a request.

    リクエストを生成するときに有効な URL が必要です。
    """


class TooManyRedirects(RequestException):
    """
    .. Too many redirects.

    リダイレクト回数が多すぎる時に送出されます。
    """


class MissingSchema(RequestException, ValueError):
    """The URL schema (e.g. http or https) is missing."""


class InvalidSchema(RequestException, ValueError):
    """See defaults.py for valid schemas."""


class InvalidURL(RequestException, ValueError):
    """The URL provided was somehow invalid."""


class InvalidHeader(RequestException, ValueError):
    """The header value provided was somehow invalid."""


class ChunkedEncodingError(RequestException):
    """The server declared chunked encoding but sent an invalid chunk."""


class ContentDecodingError(RequestException, BaseHTTPError):
    """Failed to decode response content"""


class StreamConsumedError(RequestException, TypeError):
    """The content for this response was already consumed"""


class RetryError(RequestException):
    """Custom retries logic failed"""


class UnrewindableBodyError(RequestException):
    """Requests encountered an error when trying to rewind a body"""

# Warnings


class RequestsWarning(Warning):
    """Base warning for Requests."""
    pass


class FileModeWarning(RequestsWarning, DeprecationWarning):
    """A file was opened in text mode, but Requests determined its binary length."""
    pass


class RequestsDependencyWarning(RequestsWarning):
    """An imported dependency doesn't match the expected version range."""
    pass


requests/adapters.py
# -*- coding: utf-8 -*-

"""
requests.adapters
~~~~~~~~~~~~~~~~~

This module contains the transport adapters that Requests uses to define
and maintain connections.
"""

import os.path
import socket

from urllib3.poolmanager import PoolManager, proxy_from_url
from urllib3.response import HTTPResponse
from urllib3.util import Timeout as TimeoutSauce
from urllib3.util.retry import Retry
from urllib3.exceptions import ClosedPoolError
from urllib3.exceptions import ConnectTimeoutError
from urllib3.exceptions import HTTPError as _HTTPError
from urllib3.exceptions import MaxRetryError
from urllib3.exceptions import NewConnectionError
from urllib3.exceptions import ProxyError as _ProxyError
from urllib3.exceptions import ProtocolError
from urllib3.exceptions import ReadTimeoutError
from urllib3.exceptions import SSLError as _SSLError
from urllib3.exceptions import ResponseError

from .models import Response
from .compat import urlparse, basestring
from .utils import (DEFAULT_CA_BUNDLE_PATH, get_encoding_from_headers,
                    prepend_scheme_if_needed, get_auth_from_url, urldefragauth,
                    select_proxy)
from .structures import CaseInsensitiveDict
from .cookies import extract_cookies_to_jar
from .exceptions import (ConnectionError, ConnectTimeout, ReadTimeout, SSLError,
                         ProxyError, RetryError, InvalidSchema)
from .auth import _basic_auth_str

try:
    from urllib3.contrib.socks import SOCKSProxyManager
except ImportError:
    def SOCKSProxyManager(*args, **kwargs):
        raise InvalidSchema("Missing dependencies for SOCKS support.")

DEFAULT_POOLBLOCK = False
DEFAULT_POOLSIZE = 10
DEFAULT_RETRIES = 0
DEFAULT_POOL_TIMEOUT = None


class BaseAdapter(object):
    """
    .. The Base Transport Adapter

    ベースのトランスポートアダプタ。
    """

    def __init__(self):
        super(BaseAdapter, self).__init__()

    def send(self, request, stream=False, timeout=None, verify=True,
             cert=None, proxies=None):
        """
        .. Sends PreparedRequest object. Returns Response object.

        PreparedRequest オブジェクトを送信します。Response オブジェクトを返却します。

        .. :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
        .. :param stream: (optional) Whether to stream the request content.
        .. :param timeout: (optional) How long to wait for the server to send
        ..     data before giving up, as a float, or a :ref:`(connect timeout,
        ..     read timeout) <timeouts>` tuple.
        .. :type timeout: float or tuple
        .. :param verify: (optional) Either a boolean, in which case it controls whether we verify
        ..     the server's TLS certificate, or a string, in which case it must be a path
        ..     to a CA bundle to use
        .. :param cert: (optional) Any user-provided SSL certificate to be trusted.
        .. :param proxies: (optional) The proxies dictionary to apply to the request.

        :param request: :class:`PreparedRequest <PreparedRequest>` が送信されます。
        :param stream: (任意) リクエストのコンテンツをストリーミングするかどうか。
        :param timeout: (任意) サーバーからのデータ返却をどれくらい待つかを float か :ref:`(connect timeout, read timeout) <timeouts>` のタプルで指定します。
        :type timeout: float or tuple
        :param verify: (optional) Either a boolean, in which case it controls whether we verify
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use
        :param cert: (任意) 信頼できるユーザーが提供している SSL 証明書。
        :param proxies: (任意) リクエストに付与するプロキシのディクショナリ。
        """
        raise NotImplementedError

    def close(self):
        """
        .. Cleans up adapter specific items.

        アダプタ特有の項目の後片付けをします。
        """
        raise NotImplementedError


class HTTPAdapter(BaseAdapter):
    """
    .. The built-in HTTP Adapter for urllib3.

    urllib3 用の組み込みの HTTP アダプタ。

    .. Provides a general-case interface for Requests sessions to contact HTTP and
       HTTPS urls by implementing the Transport Adapter interface. This class will
       usually be created by the :class:`Session <Session>` class under the
       covers.

    Requests のセッションがトランスポートアダプタのインターフェースを実装することによって、
    HTTP と HTTPS の URL に接続するための一般的なユースケースで使うインターフェースを提供しています。
    このクラスは通常、:class:`Session <Session>` クラスと使うために作成されます。

    .. :param pool_connections: The number of urllib3 connection pools to cache.
    .. :param pool_maxsize: The maximum number of connections to save in the pool.
    .. :param max_retries: The maximum number of retries each connection
    ..     should attempt. Note, this applies only to failed DNS lookups, socket
    ..     connections and connection timeouts, never to requests where data has
    ..     made it to the server. By default, Requests does not retry failed
    ..     connections. If you need granular control over the conditions under
    ..     which we retry a request, import urllib3's ``Retry`` class and pass
    ..     that instead.
    .. :param pool_block: Whether the connection pool should block for connections.

    :param pool_connections: キャッシュする urllib3 のコネクションプールの数。
    :param pool_maxsize: コネクションプールに保存する接続の最大数。
    :param max_retries: それぞれの接続が試行する最大回数。
        これは、失敗した DNS ルックアップ、ソケット接続、コネクションのタイムアウトのみ適用され、
        サーバーに送信したリクエストのデータには適用されません。
        デフォルトでは、Requests は失敗した接続の再試行をしません。
        リクエストを再試行する条件を細かく制御する必要がある場合、urllib3 の ``Retry``
        クラスをインポートし、代わりとして渡します。
    :param pool_block: コネクションプールが接続をブロックするかどうか。

    .. Usage

    使い方::

      >>> import requests
      >>> s = requests.Session()
      >>> a = requests.adapters.HTTPAdapter(max_retries=3)
      >>> s.mount('http://', a)
    """
    __attrs__ = ['max_retries', 'config', '_pool_connections', '_pool_maxsize',
                 '_pool_block']

    def __init__(self, pool_connections=DEFAULT_POOLSIZE,
                 pool_maxsize=DEFAULT_POOLSIZE, max_retries=DEFAULT_RETRIES,
                 pool_block=DEFAULT_POOLBLOCK):
        if max_retries == DEFAULT_RETRIES:
            self.max_retries = Retry(0, read=False)
        else:
            self.max_retries = Retry.from_int(max_retries)
        self.config = {}
        self.proxy_manager = {}

        super(HTTPAdapter, self).__init__()

        self._pool_connections = pool_connections
        self._pool_maxsize = pool_maxsize
        self._pool_block = pool_block

        self.init_poolmanager(pool_connections, pool_maxsize, block=pool_block)

    def __getstate__(self):
        return dict((attr, getattr(self, attr, None)) for attr in
                    self.__attrs__)

    def __setstate__(self, state):
        # Can't handle by adding 'proxy_manager' to self.__attrs__ because
        # self.poolmanager uses a lambda function, which isn't pickleable.
        self.proxy_manager = {}
        self.config = {}

        for attr, value in state.items():
            setattr(self, attr, value)

        self.init_poolmanager(self._pool_connections, self._pool_maxsize,
                              block=self._pool_block)

    def init_poolmanager(self, connections, maxsize, block=DEFAULT_POOLBLOCK, **pool_kwargs):
        """Initializes a urllib3 PoolManager.

        This method should not be called from user code, and is only
        exposed for use when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        :param connections: The number of urllib3 connection pools to cache.
        :param maxsize: The maximum number of connections to save in the pool.
        :param block: Block when no free connections are available.
        :param pool_kwargs: Extra keyword arguments used to initialize the Pool Manager.
        """
        # save these values for pickling
        self._pool_connections = connections
        self._pool_maxsize = maxsize
        self._pool_block = block

        self.poolmanager = PoolManager(num_pools=connections, maxsize=maxsize,
                                       block=block, strict=True, **pool_kwargs)

    def proxy_manager_for(self, proxy, **proxy_kwargs):
        """Return urllib3 ProxyManager for the given proxy.

        This method should not be called from user code, and is only
        exposed for use when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        :param proxy: The proxy to return a urllib3 ProxyManager for.
        :param proxy_kwargs: Extra keyword arguments used to configure the Proxy Manager.
        :returns: ProxyManager
        :rtype: urllib3.ProxyManager
        """
        if proxy in self.proxy_manager:
            manager = self.proxy_manager[proxy]
        elif proxy.lower().startswith('socks'):
            username, password = get_auth_from_url(proxy)
            manager = self.proxy_manager[proxy] = SOCKSProxyManager(
                proxy,
                username=username,
                password=password,
                num_pools=self._pool_connections,
                maxsize=self._pool_maxsize,
                block=self._pool_block,
                **proxy_kwargs
            )
        else:
            proxy_headers = self.proxy_headers(proxy)
            manager = self.proxy_manager[proxy] = proxy_from_url(
                proxy,
                proxy_headers=proxy_headers,
                num_pools=self._pool_connections,
                maxsize=self._pool_maxsize,
                block=self._pool_block,
                **proxy_kwargs)

        return manager

    def cert_verify(self, conn, url, verify, cert):
        """
        .. Verify a SSL certificate. This method should not be called from user
           code, and is only exposed for use when subclassing the
           :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        SSL 証明書を検証します。
        このメソッドは、使用者のコードから呼び出されるべきではなくて、
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>` をサブクラス化して使う時のみ公開されます。

        .. :param conn: The urllib3 connection object associated with the cert.
        .. :param url: The requested URL.
        .. :param verify: Either a boolean, in which case it controls whether we verify
        ..     the server's TLS certificate, or a string, in which case it must be a path
        ..     to a CA bundle to use
        .. :param cert: The SSL certificate to verify.

        :param conn: 証明書に関連付けられた urllib3 のコネクションオブジェクト。
        :param url: リクエストした URL 。
        :param verify: Either a boolean, in which case it controls whether we verify
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use
        :param cert: 検証する SSL 証明書。
        """
        if url.lower().startswith('https') and verify:

            cert_loc = None

            # Allow self-specified cert location.
            if verify is not True:
                cert_loc = verify

            if not cert_loc:
                cert_loc = DEFAULT_CA_BUNDLE_PATH

            if not cert_loc or not os.path.exists(cert_loc):
                raise IOError("Could not find a suitable TLS CA certificate bundle, "
                              "invalid path: {0}".format(cert_loc))

            conn.cert_reqs = 'CERT_REQUIRED'

            if not os.path.isdir(cert_loc):
                conn.ca_certs = cert_loc
            else:
                conn.ca_cert_dir = cert_loc
        else:
            conn.cert_reqs = 'CERT_NONE'
            conn.ca_certs = None
            conn.ca_cert_dir = None

        if cert:
            if not isinstance(cert, basestring):
                conn.cert_file = cert[0]
                conn.key_file = cert[1]
            else:
                conn.cert_file = cert
                conn.key_file = None
            if conn.cert_file and not os.path.exists(conn.cert_file):
                raise IOError("Could not find the TLS certificate file, "
                              "invalid path: {0}".format(conn.cert_file))
            if conn.key_file and not os.path.exists(conn.key_file):
                raise IOError("Could not find the TLS key file, "
                              "invalid path: {0}".format(conn.key_file))

    def build_response(self, req, resp):
        """
        .. Builds a :class:`Response <requests.Response>` object from a urllib3
           response. This should not be called from user code, and is only exposed
           for use when subclassing the
           :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`

        urllib3 のレスポンスから :class:`Response <requests.Response>` オブジェクトを構築します。
        これは、使用者のコードから呼び出されるべきではなくて、:class:`HTTPAdapter <requests.adapters.HTTPAdapter>` をサブクラス化して使う時のみ公開されます。

        .. :param req: The :class:`PreparedRequest <PreparedRequest>` used to generate the response.
        .. :param resp: The urllib3 response object.
        .. :rtype: requests.Response

        :param req: レスポンスを生成するための :class:`PreparedRequest <PreparedRequest>` 。
        :param resp: urllib3 レスポンスオブジェクト。
        :rtype: requests.Response
        """
        response = Response()

        # Fallback to None if there's no status_code, for whatever reason.
        response.status_code = getattr(resp, 'status', None)

        # Make headers case-insensitive.
        response.headers = CaseInsensitiveDict(getattr(resp, 'headers', {}))

        # Set encoding.
        response.encoding = get_encoding_from_headers(response.headers)
        response.raw = resp
        response.reason = response.raw.reason

        if isinstance(req.url, bytes):
            response.url = req.url.decode('utf-8')
        else:
            response.url = req.url

        # Add new cookies from the server.
        extract_cookies_to_jar(response.cookies, req, resp)

        # Give the Response some context.
        response.request = req
        response.connection = self

        return response

    def get_connection(self, url, proxies=None):
        """Returns a urllib3 connection for the given URL. This should not be
        called from user code, and is only exposed for use when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        :param url: The URL to connect to.
        :param proxies: (optional) A Requests-style dictionary of proxies used on this request.
        :rtype: urllib3.ConnectionPool
        """
        proxy = select_proxy(url, proxies)

        if proxy:
            proxy = prepend_scheme_if_needed(proxy, 'http')
            proxy_manager = self.proxy_manager_for(proxy)
            conn = proxy_manager.connection_from_url(url)
        else:
            # Only scheme should be lower case
            parsed = urlparse(url)
            url = parsed.geturl()
            conn = self.poolmanager.connection_from_url(url)

        return conn

    def close(self):
        """Disposes of any internal state.

        Currently, this closes the PoolManager and any active ProxyManager,
        which closes any pooled connections.
        """
        self.poolmanager.clear()
        for proxy in self.proxy_manager.values():
            proxy.clear()

    def request_url(self, request, proxies):
        """Obtain the url to use when making the final request.

        If the message is being sent through a HTTP proxy, the full URL has to
        be used. Otherwise, we should only use the path portion of the URL.

        This should not be called from user code, and is only exposed for use
        when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
        :param proxies: A dictionary of schemes or schemes and hosts to proxy URLs.
        :rtype: str
        """
        proxy = select_proxy(request.url, proxies)
        scheme = urlparse(request.url).scheme

        is_proxied_http_request = (proxy and scheme != 'https')
        using_socks_proxy = False
        if proxy:
            proxy_scheme = urlparse(proxy).scheme.lower()
            using_socks_proxy = proxy_scheme.startswith('socks')

        url = request.path_url
        if is_proxied_http_request and not using_socks_proxy:
            url = urldefragauth(request.url)

        return url

    def add_headers(self, request, **kwargs):
        """
        .. Add any headers needed by the connection. As of v2.0 this does
           nothing by default, but is left for overriding by users that subclass
           the :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        接続に必要なヘッダーを追加します。
        v2.0 では、デフォルトでは何もしませんが、
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>` をサブクラスとして、
        上書きするために残されています。

        .. This should not be called from user code, and is only exposed for use
           when subclassing the
           :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        これは使用者のコードから呼び出されるべきではなくて、
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>` をサブクラス化して使う時のみ公開されます。

        .. :param request: The :class:`PreparedRequest <PreparedRequest>` to add headers to.
        .. :param kwargs: The keyword arguments from the call to send().

        :param request: ヘッダーに追加するための :class:`PreparedRequest <PreparedRequest>`。
        :param kwargs: send() から呼び出された時に渡ってくるキーワード引数。
        """
        pass

    def proxy_headers(self, proxy):
        """Returns a dictionary of the headers to add to any request sent
        through a proxy. This works with urllib3 magic to ensure that they are
        correctly sent to the proxy, rather than in a tunnelled request if
        CONNECT is being used.

        This should not be called from user code, and is only exposed for use
        when subclassing the
        :class:`HTTPAdapter <requests.adapters.HTTPAdapter>`.

        :param proxies: The url of the proxy being used for this request.
        :rtype: dict
        """
        headers = {}
        username, password = get_auth_from_url(proxy)

        if username:
            headers['Proxy-Authorization'] = _basic_auth_str(username,
                                                             password)

        return headers

    def send(self, request, stream=False, timeout=None, verify=True, cert=None, proxies=None):
        """Sends PreparedRequest object. Returns Response object.

        :param request: The :class:`PreparedRequest <PreparedRequest>` being sent.
        :param stream: (optional) Whether to stream the request content.
        :param timeout: (optional) How long to wait for the server to send
            data before giving up, as a float, or a :ref:`(connect timeout,
            read timeout) <timeouts>` tuple.
        :type timeout: float or tuple or urllib3 Timeout object
        :param verify: (optional) Either a boolean, in which case it controls whether
            we verify the server's TLS certificate, or a string, in which case it
            must be a path to a CA bundle to use
        :param cert: (optional) Any user-provided SSL certificate to be trusted.
        :param proxies: (optional) The proxies dictionary to apply to the request.
        :rtype: requests.Response
        """

        conn = self.get_connection(request.url, proxies)

        self.cert_verify(conn, request.url, verify, cert)
        url = self.request_url(request, proxies)
        self.add_headers(request)

        chunked = not (request.body is None or 'Content-Length' in request.headers)

        if isinstance(timeout, tuple):
            try:
                connect, read = timeout
                timeout = TimeoutSauce(connect=connect, read=read)
            except ValueError as e:
                # this may raise a string formatting error.
                err = ("Invalid timeout {0}. Pass a (connect, read) "
                       "timeout tuple, or a single float to set "
                       "both timeouts to the same value".format(timeout))
                raise ValueError(err)
        elif isinstance(timeout, TimeoutSauce):
            pass
        else:
            timeout = TimeoutSauce(connect=timeout, read=timeout)

        try:
            if not chunked:
                resp = conn.urlopen(
                    method=request.method,
                    url=url,
                    body=request.body,
                    headers=request.headers,
                    redirect=False,
                    assert_same_host=False,
                    preload_content=False,
                    decode_content=False,
                    retries=self.max_retries,
                    timeout=timeout
                )

            # Send the request.
            else:
                if hasattr(conn, 'proxy_pool'):
                    conn = conn.proxy_pool

                low_conn = conn._get_conn(timeout=DEFAULT_POOL_TIMEOUT)

                try:
                    low_conn.putrequest(request.method,
                                        url,
                                        skip_accept_encoding=True)

                    for header, value in request.headers.items():
                        low_conn.putheader(header, value)

                    low_conn.endheaders()

                    for i in request.body:
                        low_conn.send(hex(len(i))[2:].encode('utf-8'))
                        low_conn.send(b'\r\n')
                        low_conn.send(i)
                        low_conn.send(b'\r\n')
                    low_conn.send(b'0\r\n\r\n')

                    # Receive the response from the server
                    try:
                        # For Python 2.7+ versions, use buffering of HTTP
                        # responses
                        r = low_conn.getresponse(buffering=True)
                    except TypeError:
                        # For compatibility with Python 2.6 versions and back
                        r = low_conn.getresponse()

                    resp = HTTPResponse.from_httplib(
                        r,
                        pool=conn,
                        connection=low_conn,
                        preload_content=False,
                        decode_content=False
                    )
                except:
                    # If we hit any problems here, clean up the connection.
                    # Then, reraise so that we can handle the actual exception.
                    low_conn.close()
                    raise

        except (ProtocolError, socket.error) as err:
            raise ConnectionError(err, request=request)

        except MaxRetryError as e:
            if isinstance(e.reason, ConnectTimeoutError):
                # TODO: Remove this in 3.0.0: see #2811
                if not isinstance(e.reason, NewConnectionError):
                    raise ConnectTimeout(e, request=request)

            if isinstance(e.reason, ResponseError):
                raise RetryError(e, request=request)

            if isinstance(e.reason, _ProxyError):
                raise ProxyError(e, request=request)

            if isinstance(e.reason, _SSLError):
                # This branch is for urllib3 v1.22 and later.
                raise SSLError(e, request=request)

            raise ConnectionError(e, request=request)

        except ClosedPoolError as e:
            raise ConnectionError(e, request=request)

        except _ProxyError as e:
            raise ProxyError(e)

        except (_SSLError, _HTTPError) as e:
            if isinstance(e, _SSLError):
                # This branch is for urllib3 versions earlier than v1.22
                raise SSLError(e, request=request)
            elif isinstance(e, ReadTimeoutError):
                raise ReadTimeout(e, request=request)
            else:
                raise

        return self.build_response(request, resp)
